packages/flutter/lib/src/cupertino/search_field.dart

```diff
@@ -0,0 +1,412 @@
+    }
+  }
+
+  @override
+  void didUpdateWidget(CupertinoSearchTextField oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (widget.controller == null && oldWidget.controller != null) {
+      _createLocalController(oldWidget.controller!.value);
+    } else if (widget.controller != null && oldWidget.controller == null) {
+      unregisterFromRestoration(_controller!);
+      _controller!.dispose();
+      _controller = null;
+    }
+  }
+
+  @override
+  void restoreState(RestorationBucket? oldBucket, bool initialRestore) {
+    if (_controller != null) {
+      _registerController();
+    }
+  }
+
+  void _registerController() {
+    assert(_controller != null);
+    registerForRestoration(_controller!, 'controller');
+    _controller!.value.addListener(updateKeepAlive);
+  }
+
+  void _createLocalController([TextEditingValue? value]) {
+    assert(_controller == null);
+    _controller = value == null
+        ? RestorableTextEditingController()
+        : RestorableTextEditingController.fromValue(value);
+    if (!restorePending) {
+      _registerController();
+    }
+  }
+
+  @override
+  String? get restorationId => widget.restorationId;
+
+  @override
+  bool get wantKeepAlive => _controller?.value.text.isNotEmpty == true;
+
+  @override
+  Widget build(BuildContext context) {
+    super.build(context); // See AutomaticKeepAliveClientMixin.
+
+    final String placeholder = widget.placeholder ??
+        CupertinoLocalizations.of(context)?.searchTextFieldPlaceholerLabel ??
+        'Search';
+
+    final TextStyle placeholderStyle = widget.placeholderStyle ??
+        TextStyle(
+          color: CupertinoDynamicColor.resolve(
+              CupertinoColors.secondaryLabel, context),
+        );
+
+    // The icon size will be scaled by a factor of the accessibility text scale,
+    // to follow the behavior of `UISearchTextField`.
+    final double scaledIconSize =
+        MediaQuery.textScaleFactorOf(context) * widget.itemSize;
+
+    // If decoration was not provided, create a decoration with the provided
+    // background color and border radius.
+    final BoxDecoration decoration = widget.decoration ??
+        BoxDecoration(
+          color: CupertinoDynamicColor.resolve(
+              widget.backgroundColor ?? CupertinoColors.tertiarySystemFill,
+              context),
+          borderRadius: widget.borderRadius ?? _kDefaultBorderRadius,
+        );
+
+    final IconThemeData iconThemeData = IconThemeData(
+        color: CupertinoDynamicColor.resolve(widget.itemColor, context),
+        size: scaledIconSize);
+
+    final Widget prefix = Padding(
+      child: IconTheme(
+          child: const Icon(CupertinoIcons.search), data: iconThemeData),
+      padding: widget.prefixInsets,
+    );
+
+    final Widget suffix = Padding(
+      child: CupertinoButton(
+        child: IconTheme(child: widget.suffixIcon, data: iconThemeData),
+        onPressed: widget.onSuffixTap ??
+            () {
```

Comment:
I'd put this in an instance method instead of a closure here. The difference is minute but `build` is called more than once throughout the lifetime of this widget instance. Each build invocation would create a new closure.